### PR TITLE
Target client improvements

### LINKED ISSUE #92: Add option to disable target cache
**How to categorize this issue?**

/area quality robustness cost
/kind bug
/priority normal

**What happened**:

We observed a Shoot cluster which had a lot of very large secrets (~3000 with ~1MiB each).
As the gardener-resource-manager caches all objects in the target cluster it touches, its cache had grown to an inacceptable size:
```
$ k top po
NAME                                          CPU(cores)   MEMORY(bytes)
gardener-resource-manager-77876b69b7-kvvkz    400m         5875Mi
```

Also, one of its worker routines trying to get a secret for applying was stuck, as the client tried to aquire a watch, but the API server panic'd (for some reasons), so the client waited forever for the watch cache to sync:
```
I1127 13:55:38.982551       1 trace.go:201] Trace[4834220]: "Reflector ListAndWatch" name:sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:224 (27-Nov-2020 13:54:00.969) (total time: 60013ms):
Trace[4834220]: [1m0.013079589s] [1m0.013079589s] END
E1127 13:55:38.982571       1 reflector.go:178] sigs.k8s.io/controller-runtime/pkg/cache/internal/informers_map.go:224: Failed to list *v1.Secret: stream error when reading response body, may be caused by closed connection. Please retry. Original error: stream error: stream ID 57251; INTERNAL_ERROR
```

**What you expected to happen**:

grm should be resistent to such user behaviour and should therefore allow to disable the target cache via a command line flag.
This way, gardener can deploy grm instances for Shoots without a cache for the Shoot API server to avoid failing watches and minimize the memory footprint.

/cc @mandelsoft @timuthy 

**Environment**:

- Gardener-Resource-Manager version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### LINKED ISSUE #93: Switch to DynamicRESTMapper for target client
**How to categorize this issue?**

/area scalability
/kind cleanup
/priority normal

**What would you like to be added**:

With the recent improvements in controller-runtime (https://github.com/kubernetes-sigs/controller-runtime/pull/1151), we can/should replace the `DeferredDiscoveryRESTMapper` by a `DynamicRESTMapper` to avoid explicit rediscovery.

https://github.com/gardener/gardener-resource-manager/blob/827ae66273760c834b2dcf34459d8cf2fca5a387/pkg/controller/managedresources/controller.go#L463-L468

This way, grm immediately rediscovers resources if some of them are not known (yet), so that no extra reconciliation is needed.
It will still adhere to the specified rate limits, so we don't do unnecessary extraneous discovery calls (even across managed resources - which wasn't the case before).

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area scalability robustness quality
/kind enhancement
/priority normal

**What this PR does / why we need it**:

This PR improves the following aspects of the client for the target cluster:
- The cache can be disabled via the newly added command line flag `--target-disable-cache`. In this case, grm won't cache any objects of the target cluster and always talk to the target API server directly (fixes #92).
- Switches to a `DynamicRESTMapper`, which will ensure that resources are rediscovered on `NoMatchErrors`, but discovery calls are not executed more than once every 10 minutes (fixes #93).

I took the chance to refactor the contents of [`app.go`](https://github.com/gardener/gardener-resource-manager/blob/290db4bffb1f1126aa939bc71218098ff4bbd8f8/cmd/gardener-resource-manager/app/app.go#L69), as I found it to be very overloaded and hard to comprehend.
Now
- both the health and secret controller are placed in their own controller package
- controllers and the corresponding options are setup in the respective controller packages in `add.go`
- the health endpoint is moved into its own dedicated package
- the `ClassFilter` logic is moved into its own dedicated package
- all remaining general options (i.e. command line flags) are handled in dedicated files under `cmd`
- `app.go` only instruments those options and `AddToManager` functions to put everything together. This is very similar to how we construct manager, options and controllers in our extensions.

Also, I switched to our gcr copies of the base images.

**Special notes for your reviewer**:
Similar to g/g, grm now also vendors our c-r fork in order to get the fix from https://github.com/kubernetes-sigs/controller-runtime/pull/1151:
`sigs.k8s.io/controller-runtime => github.com/gardener/controller-runtime v0.6.3-gardener.1`

We can drop it, once we have upgraded to c-r@v0.7.

**Release note**:

```feature developer
The cache of the kubernetes client for the target cluster can now be disabled via the `--target-disable-cache` flag.
```
```other operator
gardener-resource-manager now uses a `DynamicRESTMapper`, which will reduce the amount of explicit discovery calls and faster reconciliation loops and some cases.
```

### CODE REVIEW

**Comment by timebertt** [comment]
Also tested this PR in my local gardener setup, shoot creation/reconciliation/hibernation/wakeup/deletion works fine.
Also did some load testing with large secrets like described in #92 and grm was completely stable and at a constant level of `~30MiB` of Memory 💰

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice PR. Now it looks very similar to what we do in the Gardener extension which will further simplify maintenance, thanks for that. I only found one nit for now.

File: pkg/cmd/manager.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+	leaderElection              bool
+	leaderElectionNamespace     string
+	leaderElectionLeaseDuration time.Duration
+	leaderElectionRenewDeadline time.Duration
+	leaderElectionRetryPeriod   time.Duration
+
+	completed *ManagerConfig
+}
+
+type ManagerConfig struct {
```
[the remark this answers is not in the document]
Comment (timebertt) [reply pkg/cmd/manager.go:41]: Sure thing, done :)

**Review by timuthy — APPROVED** [review]
/lgtm

**Pull request merged**